float8_experimental/float8_dynamic_linear.py

```diff
@@ -95,7 +113,110 @@ def from_float(cls, mod, emulate: bool = False) -> "Float8DynamicLinear":
                 "bias": False,
             }
             new_mod = cls(**super_kwargs)
-        new_mod.weight = mod.weight
+        new_mod.weight = (
+            nn.Parameter(Float8DynamicLinearWeightTensor(mod.weight, emulate))
+            if config.enable_fsdp_fp8_all_gather
+            else mod.weight
+        )
         new_mod.bias = mod.bias
         new_mod.emulate = emulate
         return new_mod
+
+
+# FSDP pads its local tensor on dim-0. The subclass should be preserved such
+# that the padded local tensor (and any transformations like copying to GPU)
+# is of the subclass as well.
+_ops_to_preserve_subclass = {
+    torch.ops.aten.empty_like.default,
+    torch.ops.aten.new_zeros.default,
+    torch.ops.aten.slice.Tensor,
+    torch.ops.aten.copy_.default,
+    torch.ops.aten.view.default,
+    torch.ops.aten.as_strided.default,
+    torch.ops.aten._to_copy.default,
+    torch.ops.aten._pin_memory.default,
+}
+
+
+class Float8DynamicLinearWeightTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, tensor: torch.Tensor, emulate: bool):
+        return torch.Tensor._make_wrapper_subclass(
+            cls,
+            tensor.size(),
+            strides=tensor.stride(),
+            storage_offset=tensor.storage_offset(),
+            memory_format=suggest_memory_format(tensor),
+            dtype=tensor.dtype,
+            layout=tensor.layout,
+            device=tensor.device,
+            pin_memory=tensor.is_pinned(),
+            requires_grad=tensor.requires_grad,
+        )
+
+    def __init__(self, tensor: torch.Tensor, emulate: bool):
+        self._tensor = tensor
+        self._emulate = emulate
+        # Optional cache for pre-computed fp8 data/scale
+        self._fp8_data: Optional[torch.Tensor] = None
```

Comment:
just curious, what's the intended usage of this?  Is there a risk for the cache to become invalid?

Reply:
This is so that we can pre-compute all of the `Float8DynamicLinear` modules' fp8 casts together to only issue one all-reduce and to fuse some computation.

The cache can become invalid if the user does not call the pre-compute function after every optimizer step.

Reply:
is pre-compute called by users or is it handled by other parts of the system?

Reply:
I moved this to the next PR (which is WIP), but the pre-compute would be called by the user. It would be similar to the sync amax function for delayed scaling. Users would need to call it after optimizer step and before the next forward.